query/common/src/main/java/org/codice/ddf/admin/common/PrioritizedBatchExecutor.java

```diff
@@ -0,0 +1,242 @@
+
+    /**
+     * Start task execution and blocks until the highest priority task batch has returned a valid result
+     * according to the task handler, then cleans up remaining tasks. The current instance of the
+     * {@code PrioritizedBatchExecutor} is not useable after calling {@code getFirst()}.
+     *
+     * @return an {@code Optional} containing a task's result, if there was one
+     */
+    public Optional<T> getFirst() {
+        try {
+            List<CompletionService<T>> prioritizedCompletionServices =
+                    getPrioritizedCompletionServices();
+
+            for (int i = 0; i < tasks.size(); i++) {
+                List<Callable<T>> taskBatch = tasks.get(i);
+                CompletionService<T> completionService = prioritizedCompletionServices.get(i);
+
+                LOGGER.debug("Executing batch {} of {}.", i + 1, tasks.size());
+                for (int j = 0; j < taskBatch.size(); j++) {
+                    try {
+                        LOGGER.debug("\tAttempting to take from completion service queue.");
+                        Future<T> taskFuture = completionService.take();
+
+                        Optional<T> result = handleTaskResult(taskFuture);
+                        if (result.isPresent()) {
+                            LOGGER.debug("\tReturning valid task result {} of {} tasks.",
+                                    j + 1,
+                                    taskBatch.size());
+                            return result;
+                        }
+                    } catch (InterruptedException e) {
+                        LOGGER.debug("\t\tFailed to get future from completion service.");
+                    }
+
+                    LOGGER.debug("\tReceived invalid task result {} of {} tasks.",
+                            j + 1,
+                            taskBatch.size());
+                }
+                LOGGER.debug("\tFound no valid results in batch {}.", i + 1);
+            }
+
+            LOGGER.debug("Found no valid responses in all {} batches.", tasks.size());
+            return Optional.empty();
+        } finally {
+            cleanUp();
+        }
+    }
+
+    /**
+     * Start task execution and blocks until the highest priority task batch has returned a valid result
+     * according to the task handler, then cleans up remaining tasks. The current instance of the
+     * {@code PrioritizedBatchExecutor} is not useable after calling {@code getFirst(long, TimeUnit)}.
+     *
+     * @param batchWaitTime amount of time to wait for each batch execution
+     * @param timeUnit      {@code TimeUnit} to use for the {@code batchWaitTime}
+     * @return an {@code Optional} containing a task's result, if there was one
+     */
+    public Optional<T> getFirst(long batchWaitTime, TimeUnit timeUnit) {
+        if (batchWaitTime < 0) {
+            throw new IllegalArgumentException("Batch wait time must be 0 or greater.");
+        }
+
+        if (batchWaitTime == 0) {
```

Comment:
✏️ If the wait time is greater than zero, you'll want to validate that `timeUnit` is non-null.